src/Streamly/Internal/Data/Stream/Async/Channel.hs

```diff
@@ -190,9 +191,15 @@ mkEnqueue :: MonadAsync m =>
 mkEnqueue chan runner = do
     runInIO <- askRunInIO
     return
-        $ let q stream =
+        $ let q stream = do
                 -- Enqueue the outer loop
                 liftIO $ enqueue chan False (runInIO, runner q stream)
+                -- XXX In case of eager dispatch we can just directly dispatch
+                -- a worker with the tail stream here rather than first queuing
+                -- and then dispatching a worker which dequeues the work. The
+                -- older implementation did a direct dispatch here and its perf
+                -- characterstics looked much better.
```

Comment:
Can you elaborate a little more? Doesn't `dispatchWorker` always dequeue from the `Channel`? For directly dispatching a worker with a `Stream` we would need a new function right? That function would do what a `workLoop` does bust instead of dequing it would take the stream as a parameter?

Reply:
Yes. Currently we dispatch a worker i.e. fork a new thread, the thread runs and tries to dequeue from the channel work queue, if it finds the queue empty then the thread exits.

Here we already have the work item that we are queuing, so instead of queuing it we can pass it to the worker we are forking. So we will need a variant of dispatchWorker that would fork the workloop with the stream argument.